File: python/008-trees/main.py
FILE_NAME = "input.txt"


def process_input(input):
    result = 0
    for i, row in enumerate(input):
        for j, _ in enumerate(row):
            tmp = heigh_value(i, j, input, len(input))
            result = tmp if tmp > result else result
    return result

def process_input1(input):
    result = 0
    for i, row in enumerate(input):
        for j, _ in enumerate(row):
            tmp = heigh_value1(i, j, input, len(input))
            result += 1 if tmp else 0
    return result


def heigh_value1(row, column, input, max_size):
    directions = ["r", "l", "u", "d"]
    for i in range(1, max_size):
        tmp_directions = directions.copy()
        for direction in tmp_directions:
            if direction == "r":
                if column + i == max_size:
                    return True
                elif input[row][column] <= input[row][column + i]:
                    directions.remove(direction)
            if direction == "l":
                if column - i < 0:
                    return True
                elif input[row][column] <= input[row][column - i]:
                    directions.remove(direction)
            if direction == "d":
                if row + i == max_size:
                    return True
                elif input[row][column] <= input[row + i][column]:
                    directions.remove(direction)
            if direction == "u":
                if row - i < 0:
                    return True
                elif input[row][column] <= input[row - i][column]:
                    directions.remove(direction)
        if not len(directions):
            return False
    # print("[" + str(row) + "," + str(column) + "] -> " + str(res))
    return False

def heigh_value(row, column, input, max_size):
    directions = ["r", "l", "u", "d"]
    res = 1
    for i in range(1, max_size):
        tmp_directions = directions.copy()
        for direction in tmp_directions:
            if direction == "r":
                if column + i == max_size:
                    res *= i - 1
                    directions.remove(direction)
                elif input[row][column] <= input[row][column + i]:
                    res *= i
                    directions.remove(direction)
            if direction == "l":
                if column - i < 0:
                    res *= i - 1
                    directions.remove(direction)
                elif input[row][column] <= input[row][column - i]:
                    res *= i
                    directions.remove(direction)
            if direction == "d":
                if row + i == max_size:
                    res *= i - 1
                    directions.remove(direction)
                elif input[row][column] <= input[row + i][column]:
                    res *= i
                    directions.remove(direction)
            if direction == "u":
                if row - i < 0:
                    res *= i - 1
                    directions.remove(direction)
                elif input[row][column] <= input[row - i][column]:
                    res *= i
                    directions.remove(direction)
        if not len(directions):
            break
    # print("[" + str(row) + "," + str(column) + "] -> " + str(res))
    return res


def read_file(file_name):
    res = []
    with open(file_name, "r") as f:
        for line in f.readlines():
            res.append([int(x) for x in line.strip()])
    return res


if __name__ == "__main__":
    input = read_file(FILE_NAME)
    print(process_input1(input))

File: go/008-trees/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// func findTrees(input [][]int) int {
// 	res := 0
// 	for r, row := range input {
// 		for c, _ := range row {
// 			if isHighest2(r, c, input, len(input)) {
// 				res++
// 			}
// 		}
// 	}
// 	return res
// }

func findTrees(input [][]int) int {
	res := 0
	for r, row := range input {
		for c, tree := range row {
			score := isHighest2(r, c, input, len(input))
			fmt.Printf("[%d,%d] -> %d = %d\n", r, c, tree, score)
			if score > res {
				res = score
			}
		}
	}
	return res
}

// func isHighest2(i, j int, input [][]int, max int) bool {
// 	directions := map[string]bool{
// 		"r": true,
// 		"l": true,
// 		"t": true,
// 		"d": true,
// 	}
// 	for r := 0; r < max; r++ {
// 		for direction, _ := range directions {
// 			switch direction {
// 			case "r":
// 				if max-r-1 == j {
// 					fmt.Printf("[%d,%d] -> %d\n", i, j, input[i][j])
// 					return true
// 				}
// 				if input[i][max-r-1] >= input[i][j] {
// 					delete(directions, direction)
// 				}
// 			case "l":
// 				if r == j {
// 					fmt.Printf("[%d,%d] -> %d\n", i, j, input[i][j])
// 					return true
// 				}
// 				if input[i][r] >= input[i][j] {
// 					delete(directions, direction)
// 				}
// 			case "d":
// 				if max-r-1 == i {
// 					fmt.Printf("[%d,%d] -> %d\n", i, j, input[i][j])
// 					return true
// 				}
// 				if input[max-r-1][j] >= input[i][j] {
// 					delete(directions, direction)
// 				}
// 			case "t":
// 				if r == i {
// 					fmt.Printf("[%d,%d] -> %d\n", i, j, input[i][j])
// 					return true
// 				}
// 				if input[r][j] >= input[i][j] {
// 					delete(directions, direction)
// 				}
// 			}
// 		}
// 		if len(directions) == 0 {
// 			return false
// 		}
// 	}
// 	return false
// }

func isHighest2(i, j int, input [][]int, max int) int {
	res := 1
	directions := map[string]bool{
		"r": true,
		"l": true,
		"t": true,
		"d": true,
	}
	for r := 1; r < max-1; r++ {
		if len(directions) == 0 {
			return res
		}
		for direction, _ := range directions {
			switch direction {
			case "r":
				if j+r == max {
					res *= (r-1) 
					delete(directions, direction)
					continue
				}
				if input[i][j+r] >= input[i][j] {
					res *= r
					delete(directions, direction)
				}
			case "l":
				if j-r < 0 {
					res *= (r-1)
					delete(directions, direction)
					continue
				}
				if input[i][j-r] >= input[i][j] {
					res *= r
					delete(directions, direction)
				}
			case "d":
				if i+r == max {
					res *= (r-1)
					delete(directions, direction)
					continue
				}
				if input[i+r][j] >= input[i][j] {
					res *= r
					delete(directions, direction)
				}
			case "t":
				if i-r < 0 {
					res *= (r-1)
					delete(directions, direction)
					continue
				}
				if input[i-r][j] >= input[i][j] {
					res *= r
					delete(directions, direction)
				}
			}
		}
	}
	return res
}

func readFile(fileName string) [][]int {
	res := make([][]int, 0)
	file, _ := os.Open(fileName)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, "")
		tmp := []int{}
		for _, part := range parts {
			value, _ := strconv.Atoi(part)
			tmp = append(tmp, value)
		}
		res = append(res, tmp)
	}

	return res
}

func main() {
	input := readFile("input.txt")
	fmt.Println(findTrees(input))
	isHighest2(3, 2, input, 5)
}
